### PR TITLE
Fix tensorflow len issues

### PR DESCRIPTION
**Pull request by RandomAnass**
## Summary
To fix  TypeError issues that occur when using TensorFlow symbolic tensors with `enable_autodiff=True` in GUDHI's wasserstein distance and KNN functions.

## Problem
The issue was that `len(X)` and `len(Y)` were called on TensorFlow symbalic tensors before checking if autodiff was enabled. TensorFlow symbolic tensors don't support the `len()` function, causing a TypeError.

This is related to [issue #530](https://github.com/GUDHI/gudhi-devel/issues/530).

## Changes
1. **wasserstein.py**: Added TensorFlow tensor detection before using `len()` in the `wasserstein_distance` function
2. **knn.py**: Added similar TensorFlow tensor detection in the `fit` method

## Details
- Use `tensor.shape[0]` for TensorFlow tensors instead of `len(tensor)`
- Tried to keep backward compatibility with numpy arrays 
- Added proper error handling for missing TensorFlow dependency
- Only apply the fix when `enable_autodiff=True`

## Testing
- Both functions now handle TensorFlow symbolic tensors without throwing TypeError
- Backward compatibility with numpy arrays is maintained (from the tests i had)
- Functions work correctly when TensorFlow is not available

## Files Changed
- `src/python/gudhi/wasserstein/wasserstein.py`
- `src/python/gudhi/point_cloud/knn.py`

### CODE REVIEW

**Comment by VincentRouvreau** [comment]
Thank you for the PR !
@RandomAnass I just had a look to the Wasserstein part, and I was wondering to know if you would have an example in mind that you want to solve ?

Another way to solve it would be to move the following code (l. 353-360):
```python
    if enable_autodiff:
        import eagerpy as ep

        X_orig = ep.astensor(X)
        Y_orig = ep.astensor(Y)
        X = X_orig.numpy()
        Y = Y_orig.numpy()
```
before the call to `len(X)`.

I tried and it seems to fix #530 if we call `model.compile(optimizer='SGD', loss=wloss, run_eagerly=True)` as @mglisse was suggesting.
